File: 섹션08. A형 문제 풀이 (~2022 상반기)/04. 2048 게임 (코드트리, 2016 하반기 2번).cpp
#include <stdio.h>

#define MAX (20 + 5)

#define UP (0)
#define RIGHT (1)
#define DOWN (2)
#define LEFT (3)

int T;

int N;
int oMAP[MAX][MAX]; // original Map
int ansMAP[MAX][MAX]; // answerMap

int num_of_cases[5 + 5];

int maxAnswer;

void input()
{
	scanf("%d", &N);

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			scanf("%d", &oMAP[r][c]);
}

void printMap(int map[MAX][MAX]) // for debug
{
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= N; c++)
			printf("%2d ", map[r][c]);
		putchar('\n');
	}
	putchar('\n');
}

void copyMap(int copy[MAX][MAX], int original[MAX][MAX])
{
	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			copy[r][c] = original[r][c];
}

void printCases()
{
	for (int i = 0; i < 5; i++) printf("%d ", num_of_cases[i]);
	putchar('\n');
}

void moveUp()
{
	int tmpMAP[MAX][MAX] = { 0 };

	for (int c = 1; c <= N; c++)
	{
		int index = 1;
		for (int r = 1; r <= N; r++)
		{
			if (ansMAP[r][c] == 0) continue;

			tmpMAP[index++][c] = ansMAP[r][c];
		}

		for (int r = 1; r <= N - 1; r++)
		{
			if (tmpMAP[r][c] != tmpMAP[r + 1][c]) continue;
			if (tmpMAP[r][c] == 0) break;

			// 기준이 되는 점을 2배로 
			tmpMAP[r][c] *= 2;
			
			// 남은 칸을 앞으로 1칸씩 옮긴다. 
			for (int tr = r + 1; tr <= N - 1; tr++)
				tmpMAP[tr][c] = tmpMAP[tr + 1][c];

			tmpMAP[N][c] = 0;
		}
	}

	copyMap(ansMAP, tmpMAP);
}

void moveRight()
{
	int tmpMAP[MAX][MAX] = { 0 };

	for (int r = 1; r <= N; r++)
	{
		int index = N; // 
		for (int c = N; c >= 1; c--)
		{
			if (ansMAP[r][c] == 0) continue;

			tmpMAP[r][index--] = ansMAP[r][c];
		}

		for (int c = N; c >= 2; c--)
		{
			if (tmpMAP[r][c] != tmpMAP[r][c - 1]) continue;
			if (tmpMAP[r][c] == 0) break;

			tmpMAP[r][c] *= 2;

			for (int tc = c - 1; tc >= 2; tc--) 
				tmpMAP[r][tc] = tmpMAP[r][tc - 1];

			tmpMAP[r][1] = 0;
		}
	}

	copyMap(ansMAP, tmpMAP);
}

void moveDown() 
{
	int tmpMAP[MAX][MAX] = { 0 };

	for (int c = 1; c <= N; c++)
	{
		int index = N;
		for (int r = N; r >= 1; r--)
		{
			if (ansMAP[r][c] == 0) continue;

			tmpMAP[index--][c] = ansMAP[r][c];
		}

		for (int r = N; r >= 2; r--)
		{
			if (tmpMAP[r][c] != tmpMAP[r - 1][c]) continue;
			if (tmpMAP[r][c] == 0) break;

			tmpMAP[r][c] *= 2;
 
			for (int tr = r - 1; tr >= 2; tr--)
				tmpMAP[tr][c] = tmpMAP[tr - 1][c];

			tmpMAP[1][c] = 0;
		}
	}

	copyMap(ansMAP, tmpMAP);
}

void moveLeft()
{
	int tmpMAP[MAX][MAX] = { 0 };

	for (int r = 1; r <= N; r++)
	{
		int index = 1; // (1, 1) 부터 시작
		for (int c = 1; c <= N; c++)
		{
			if (ansMAP[r][c] == 0) continue;

			tmpMAP[r][index++] = ansMAP[r][c];
		}

		for (int c = 1; c <= N - 1; c++)
		{
			if (tmpMAP[r][c] != tmpMAP[r][c + 1]) continue;
			if (tmpMAP[r][c] == 0) break;

			tmpMAP[r][c] *= 2;
			for (int tc = c + 1; tc <= N - 1; tc++) 
				tmpMAP[r][tc] = tmpMAP[r][tc + 1];

			tmpMAP[r][N] = 0;
		}
	}

	copyMap(ansMAP, tmpMAP);
}

void simulate()
{
	copyMap(ansMAP, oMAP);

	for (int i = 0; i < 5; i++)
	{
		int direction = num_of_cases[i];

		if (direction == UP) moveUp();
		else if (direction == RIGHT) moveRight();
		else if (direction == DOWN) moveDown();
		else if (direction == LEFT) moveLeft();
	}
}

int findMax()
{
	int maxValue = 0;
	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			if (maxValue < ansMAP[r][c]) maxValue = ansMAP[r][c];

	return maxValue;
}

void DFS(int depth)
{
	if (depth == 5)
	{
		// printCases();

		simulate();

		int tmp = findMax();
		if (maxAnswer < tmp) maxAnswer = tmp;

		return;
	}

	for (int i = 0; i < 4; i++)
	{
		num_of_cases[depth] = i;

		DFS(depth + 1);
	}
}

int main()
{
	// scanf("%d", &T);
	T = 1;
	for (int tc = 1; tc <= T; tc++)
	{
		input();

		maxAnswer = 0;

		DFS(0);

		printf("%d\n", maxAnswer);
	}

	return 0;
}

File: Python/섹션08. A형 문제 풀이 (~2022 상반기)/04. 2048 게임 (코드트리, 2016 하반기 2번).py
import copy

MAX = 20 + 5

UP = 0
RIGHT = 1
DOWN = 2
LEFT = 3

oMAP = [[0] * MAX for _ in range(MAX)]
ansMAP = [[0] * MAX for _ in range(MAX)]

num_of_cases = [0] * (5 + 5)

def input_data():             
    global N
    
    N = int(input())
        
    for r in range(1, N + 1):
        oMAP[r][1:] = map(int, input().split())
                
def print_map(map): # for debug
    for r in range(N):
        print(' '.join('%2d' % num for num in map[r][1:N + 1]))
    print("")    
    
def print_cases(): 
    print(*num_of_cases[:5])
    
def move_up():
    global ansMAP
    
    tmp_map = [[0] * MAX for _ in range(MAX)]
    
    for c in range(1, N + 1):
        index = 1
        for r in range(1, N + 1):
            if ansMAP[r][c] == 0: continue
            
            tmp_map[index][c] = ansMAP[r][c]
            index += 1
            
        for r in range(1, N):
            if tmp_map[r][c] != tmp_map[r + 1][c]: continue
            if tmp_map[r][c] == 0: break
            
            # 기준이 되는 점을 2배로
            tmp_map[r][c] *= 2
            
            # 남은 칸을 앞으로 1칸식 옮긴다.
            for tr in range(r + 1, N):
                tmp_map[tr][c] = tmp_map[tr + 1][c]
            
            tmp_map[N][c] = 0
            
    ansMAP = copy.deepcopy(tmp_map)

def move_right():
    global ansMAP
    
    tmp_map = [[0] * MAX for _ in range(MAX)]
    
    for r in range(1, N + 1):
        index = N
        for c in range(N, 0, -1):
            if ansMAP[r][c] == 0: continue
            
            tmp_map[r][index] = ansMAP[r][c]
            index -= 1
            
        for c in range(N, 1, -1):
            if tmp_map[r][c] != tmp_map[r][c - 1]: continue
            if tmp_map[r][c] == 0: break
            
            tmp_map[r][c] *= 2
            
            for tc in range(c - 1, 1, -1):
                tmp_map[r][tc] = tmp_map[r][tc - 1]
            
            tmp_map[r][1] = 0
            
    ansMAP = copy.deepcopy(tmp_map)

def move_down():
    global ansMAP
    
    tmp_map = [[0] * MAX for _ in range(MAX)]
    
    for c in range(1, N + 1):
        index = N
        for r in range(N, 0, -1):
            if ansMAP[r][c] == 0: continue
            
            tmp_map[index][c] = ansMAP[r][c]
            index -= 1
            
        for r in range(N, 1, -1):
            if tmp_map[r][c] != tmp_map[r - 1][c]: continue
            if tmp_map[r][c] == 0: break
            
            tmp_map[r][c] *= 2
            
            for tr in range(r - 1, 1, -1):
                tmp_map[tr][c] = tmp_map[tr - 1][c]
            
            tmp_map[1][c] = 0
            
    ansMAP = copy.deepcopy(tmp_map)

def move_left():
    global ansMAP
    
    tmp_map = [[0] * MAX for _ in range(MAX)]
    
    for r in range(1, N + 1):
        index = 1 # (1, 1)부터 시작
        for c in range(1, N + 1):
            if ansMAP[r][c] == 0: continue
            
            tmp_map[r][index] = ansMAP[r][c]
            index += 1
            
        for c in range(1, N):
            if tmp_map[r][c] != tmp_map[r][c + 1]: continue
            if tmp_map[r][c] == 0: break
            
            tmp_map[r][c] *= 2
            
            for tc in range(c + 1, N):
                tmp_map[r][tc] = tmp_map[r][tc + 1]
            
            tmp_map[r][N] = 0
            
    ansMAP = copy.deepcopy(tmp_map)
      
def simulate():
    global ansMAP
    
    ansMAP = copy.deepcopy(oMAP)
    
    for i in range(5):
        direction = num_of_cases[i]
        
        if direction == UP: move_up()
        elif direction == RIGHT: move_right()
        elif direction == DOWN: move_down()
        elif direction == LEFT: move_left()
        
def find_max():
    return max(max(r) for r in ansMAP)

def DFS(depth):
    global max_answer
    
    if depth == 5:
        # print_cases()
        
        simulate()
        
        tmp = find_max()
        if max_answer < tmp: max_answer = tmp
        
        return
    
    for i in range(4):
        num_of_cases[depth] = i
        DFS(depth + 1)
                            
# T = int(input())
T = 1
for tc in range(T):
    global max_answer
    
    input_data()     
    
    max_answer = 0
    
    DFS(0)
    
    print(max_answer)
                     